# LGBTIQ Human Rights Scoring for Travelling -  Data Exploration
## by Thais Ruiz

## Preliminary Wrangling
This document explores a dataset of 1,777 LGBTIQ Human Rights Scoring in 14 categories on Period 2012-2021 worldwide.

In [1]:
# Import pandas and numpy, and load Gay Travel index dataset into pandas dataframe
# Data sorce from MondayMakeOver

import pandas as pd
import numpy as np

# suppress warnings from final output
import warnings
warnings.simplefilter("ignore")

df = pd.read_excel(r"C:\Downloads\GTI_2012-2021.xlsx")

pd.set_option('display.max_columns', 21)
pd.set_option('display.precision', 0)

In [2]:
df.head()

Country  Ranking  Year  Total Antidiscrimination Legislation  \
0  Afghanistan      183  2015     -9                              0   
1      Albania       55  2015      0                              2   
2      Algeria      183  2015     -9                              0   
3      Andorra       22  2015      4                              2   
4       Angola      104  2015     -4                              0   

  Marriage / Civil Partnership Adoption Allowed  LGBT Marketing  \
0                            0                0               0   
1                            0                0               0   
2                            0                0               0   
3                            1                0               0   
4                            0                0               0   

   Equal Age of Consent Religious Influence HIV Travel Restrictions  \
0                     0                  -2                       0   
1                     1                  -1                       0   
2                     0                  -2                       0   
3                     1                   0                       0   
4                     0                  -1                       0   

  Anti-Gay Laws  Homosexuality Illegal  Pride Banned Locals Hostile  \
0            -1                     -1             0             -2   
1             0                      0             0             -2   
2            -1                     -1             0             -2   
3             0                      0             0              0   
4            -1                     -1             0             -1   

   Prosecution  Murders  Death Sentence  Trans Gender Rights  \
0            0       -2              -1                  NaN   
1            0        0               0                  NaN   
2           -1       -2               0                  NaN   
3            0        0               0                  NaN   
4            0        0               0                  NaN   

   Intersex / 3rd Option  Conversion Therapy  
0                    NaN                 NaN  
1                    NaN                 NaN  
2                    NaN                 NaN  
3                    NaN                 NaN  
4                    NaN                 NaN

In [3]:
df.tail()

Country  Ranking  Year  Total  \
1772               Jamaica      134  2012    -11   
1773                Jordan      134  2012    -11   
1774                Russia      134  2012    -11   
1775  United Arab Emirates      137  2012    -12   
1776                  Iran      138  2012    -13   

     Antidiscrimination Legislation Marriage / Civil Partnership  \
1772                              0                            0   
1773                              0                            0   
1774                              0                            0   
1775                              0                            0   
1776                              0                            0   

     Adoption Allowed  LGBT Marketing  Equal Age of Consent  \
1772                0               0                     0   
1773                0               0                     1   
1774                0               0                     1   
1775                0               0                     1   
1776                0               0                     1   

     Religious Influence HIV Travel Restrictions Anti-Gay Laws  \
1772                  -1                       0            -2   
1773                  -2                      -3            -1   
1774                  -1                      -3            -1   
1775                  -1                      -2            -1   
1776                  -2                       0            -1   

      Homosexuality Illegal  Pride Banned Locals Hostile  Prosecution  \
1772                     -2             0             -2           -2   
1773                      0             0             -2           -2   
1774                      0            -2             -2           -1   
1775                     -1             0             -1           -2   
1776                     -1             0             -1           -2   

      Murders  Death Sentence  Trans Gender Rights  Intersex / 3rd Option  \
1772       -2               0                  NaN                    NaN   
1773       -2               0                  NaN                    NaN   
1774       -2               0                  NaN                    NaN   
1775        0              -5                  NaN                    NaN   
1776       -2              -5                  NaN                    NaN   

      Conversion Therapy  
1772                 NaN  
1773                 NaN  
1774                 NaN  
1775                 NaN  
1776                 NaN

In [4]:
# High-level overview of data shape and composition
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1777 entries, 0 to 1776
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country                         1777 non-null   object 
 1   Ranking                         1777 non-null   int64  
 2   Year                            1777 non-null   int64  
 3   Total                           1777 non-null   int64  
 4   Antidiscrimination Legislation  1775 non-null   object 
 5   Marriage / Civil Partnership    1776 non-null   object 
 6   Adoption Allowed                1776 non-null   object 
 7   LGBT Marketing                  1394 non-null   float64
 8   Equal Age of Consent            1777 non-null   int64  
 9   Religious Influence             1775 non-null   object 
 10  HIV Travel Restrictions         1769 non-null   object 
 11  Anti-Gay Laws                   1776 non-null   object 
 12  Homosexuality Illegal           17

In [5]:
df.sample(15)

Country  Ranking  Year  Total Antidiscrimination Legislation  \
1154              Yemen      198  2021    -15                              0   
893            Paraguay      125  2016     -5                              0   
854                Laos       79  2016     -1                              0   
62              Georgia       96  2015     -3                              1   
944            Zimbabwe      178  2016     -8                              0   
1337           Suriname      117  2020     -8                              0   
1516            Austria       15  2013      4                              1   
1430          Nicaragua       60  2014     -1                              0   
1385            Ireland       21  2014      3                              1   
1400            Reunion       30  2014      2                              1   
1722             Taiwan       74  2012     -2                              0   
122             Curacao       67  2018      0                              0   
1163  Antigua & Barbuda      107  2021     -6                              0   
1318             Poland      103  2020     -6                              1   
1232            Finland       22  2021      6                              3   

     Marriage / Civil Partnership Adoption Allowed  LGBT Marketing  \
1154                           -1                0               0   
893                             0                0               0   
854                             0                0               0   
62                              0                0               0   
944                             0                0               0   
1337                           -1                0               0   
1516                            1                0               1   
1430                            0                0               0   
1385                            1                0               0   
1400                            0                0               1   
1722                            0                0               0   
122                             0                0             NaN   
1163                           -1                0               0   
1318                           -1                0               0   
1232                            2                2               0   

      Equal Age of Consent Religious Influence HIV Travel Restrictions  \
1154                     0                  -2                      -1   
893                      0                  -1                       0   
854                      0                   0                       0   
62                       1                  -2                       0   
944                      0                  -1                       0   
1337                    -1                  -1                      -1   
1516                     1                   0                       0   
1430                     0                   0                       0   
1385                     1                   0                       0   
1400                     0                   0                       0   
1722                     0                   0                      -2   
122                      1                   0                       0   
1163                     0                   0                       0   
1318                     0                  -1                       0   
1232                     0                   0                       0   

     Anti-Gay Laws  Homosexuality Illegal  Pride Banned Locals Hostile  \
1154            -1                     -1             0             -2   
893             -1                      0             0             -1   
854              0                      0             0             -1   
62               0                      0             0             -1   
944             -1                     -1             0    

In [6]:
# Duplicates
df.duplicated().sum()

0

In [7]:
# Missingness on some categories from 2012 to 2014
Caribbean = ['Aruba', 'Belize', 'Bermuda', 'Bonaire', 'Curacao', 'Grenada', 'Guam', 'Saint Kitts & Nevis']

df[ (df.Country.isin(Caribbean)) & (df.Year.isin([2012, 2013, 2014])) ].sort_values(by = ['Country', 'Year'])

Country  Ranking  Year  Total  \
1678                Aruba       38  2012      1   
1538                Aruba       38  2013      1   
1405                Aruba       40  2014      1   
1743               Belize      104  2012     -4   
1607               Belize      105  2013     -4   
1468               Belize      105  2014     -4   
1698              Bermuda       59  2012     -1   
1561              Bermuda       60  2013     -1   
1432              Bermuda       60  2014     -1   
1686              Bonaire       47  2012      0   
1550              Bonaire       48  2013      0   
1420              Bonaire       49  2014      0   
1679              Curacao       38  2012      1   
1542              Curacao       38  2013      1   
1403              Curacao       40  2014      1   
1729              Grenada       88  2012     -3   
1592              Grenada       90  2013     -3   
1456              Grenada       90  2014     -3   
1691                 Guam       47  2012      0   
1554                 Guam       48  2013      0   
1416                 Guam       49  2014      0   
1745  Saint Kitts & Nevis      104  2012     -4   
1615  Saint Kitts & Nevis      105  2013     -4   
1475  Saint Kitts & Nevis      105  2014     -4   

     Antidiscrimination Legislation Marriage / Civil Partnership  \
1678                              0                            0   
1538                              0                            0   
1405                              0                            0   
1743                              0                            0   
1607                              0                            0   
1468                              0                            0   
1698                              0                            0   
1561                              0                            0   
1432                              0                            0   
1686                              -                            0   
1550                              -                            0   
1420                            NaN                            0   
1679                              0                            0   
1542                              0                            0   
1403                              0                            0   
1729                              0                            0   
1592                              0                            0   
1456                              0                            0   
1691                              -                            -   
1554                              -                            -   
1416                            NaN                          NaN   
1745                              0                            0   
1615                              0                            0   
1475                              0                            0   

     Adoption Allowed  LGBT Marketing  Equal Age of Consent  \
1678                0               0                     1   
1538                0               0                     1   
1405                0               0                     1   
1743                0               0                     0   
1607                0               0                     0   
1468                0               0                     0   
1698                0               0                     0   
1561                0               0                     0   
1432                0               0                     0   
1686                0               0                     1   
1550                0               0                     1   
1420                0               0                     1   
1679                0               1                     1   
1542                0               1                     1   
1403                0               1                     1   
1729                0            

In [8]:
# Change Country data type to category
df.Country = df.Country.astype("category")

# Chage Human Rights columns to numeric
factors = df.columns[4:] # Selecting factor columns (scoring categories)
df[factors] = df[factors].apply(pd.to_numeric, errors='coerce', downcast = 'integer')

In [9]:
# Country names
Countries = sorted(df.Country.unique())
print(*Countries, sep = "\n")

Afghanistan
Albania
Algeria
Andorra
Angola
Antigua & Barbuda
Argentina
Armenia
Aruba
Australia
Austria
Azerbaijan
Bahamas
Bahrain
Bangladesh
Barbados
Belarus
Belgium
Belize
Benin
Bermuda
Bhutan
Bolivia
Bonaire
Bosnia-Hercegovina
Bosnia-Herzegovina
Botswana
Brazil
British Virgin Islands
Brunei
Bulgaria
Burkina Faso
Burundi
Cabo Verde
Cambodia
Cameroon
Canada
Cape Verde
Central African Rep.
Central African Republic
Chad
Chechnya
Chile
China incl. Hong Kong
China/Hong Kong
Colombia
Comoros Islands
Cook Islands
Costa Rica
Croatia
Cuba
Curacao
Cyprus
Czech Republic
Dem. Rep. of Congo
Dem.Rep. of Congo
Denmark
Djibouti
Dominica
Dominican Republic
Ecuador
Egypt
El Salvador
Equatorial Guinea
Eritrea
Estonia
Ethiopia
Fiji
Finland
France
French Polynesia
Gabon
Gambia
Georgia
Georgia (Europe)
Germany
Ghana
Gibraltar
Greece
Greenland
Grenada
Guadeloupe
Guam
Guatemala
Guinea-Bissau
Guyana
Haiti
Honduras
Hungary
Iceland
India
Indonesia
Indonesia/Aceh
Iran
Iraq
Ireland
Israel
Italy
Ivory Coast
Jamaic

In [10]:
# Replace Country names

# Since replacing 'Virgin Islands' by 'Virgin Islands (US)', 'British Virgin Islands' become 'British Virgin Islands (US)'
# So, 'British Virgin Islands (US)' have to be replaced back to its original's

df = df.replace({'Country': {'Bosnia-Hercegovina': 'Bosnia-Herzegovina', 'Brunei': 'Brunei Darussalam',\
                             'Cabo Verde': 'Cape Verde', 'Central African Rep.': 'Central African Republic', \
                             'China/Hong Kong': 'China', 'China incl. Hong Kong': 'China', \
                             'Comoros Islands': 'Comoros',\
                             'Dem.Rep. of Congo': 'Democratic Republic of the Congo', \
                             'Dem. Rep. of Congo': 'Democratic Republic of the Congo', \
                             'Georgia (Europe)': 'Georgia', 'Macedonia': 'North Macedonia', \
                             'Maledives': 'Maldives',  \
                             'Rep. of the Congo': 'Republic of the Congo', \
                             'St. Maarten': 'Saint Martin', \
                             'USA*': 'United States', 'Virgin Islands (USA)': 'Virgin Islands (US)', \
                             'USA': 'United States', 'Virgin Islands': 'Virgin Islands (US)', \
                             'British Virgin Islands (US)': 'British Virgin Islands'}})

In [11]:
# Delete invalid Countries: Chechnya, Indonesia/Aceh, Macao

invalid_countries = df.loc[(df.Country == 'Chechnya') | (df.Country == 'Indonesia/Aceh') | (df.Country == 'Macao')]
df.drop(invalid_countries.index, inplace = True)
#invalid_countries

In [12]:
# Check if values of Total are the Sum of the scoring of Human Rights

df[ df.Total != df[factors].sum(1) ]

Country  Ranking  Year  Total  Antidiscrimination Legislation  \
1390      Brazil       21  2014      3                               1   
1408      Mexico       40  2014      1                               1   
1428  Azerbaijan       60  2014     -1                               0   
1467       India      105  2014     -4                               0   
1500        Iran      138  2014    -14                               0   
1522      Brazil       21  2013      3                               1   
1545      Mexico       38  2013      1                               1   
1560  Azerbaijan       60  2013     -1                               0   
1642      Canada        4  2012      7                               2   
1681      Mexico       38  2012      1                               1   
1705  Azerbaijan       59  2012     -1                               0   

      Marriage / Civil Partnership  Adoption Allowed  LGBT Marketing  \
1390                             2                 1               1   
1408                             0                 1               1   
1428                             0                 0               0   
1467                             0                 0               0   
1500                             0                 0               0   
1522                             2                 1               1   
1545                             0                 1               1   
1560                             0                 0               0   
1642                             2                 2               1   
1681                             0                 1               1   
1705                             0                 0               0   

      Equal Age of Consent  Religious Influence  HIV Travel Restrictions  \
1390                     1                   -1                        0   
1408                     1                   -1                        0   
1428                     0                    0                        0   
1467                     1                   -1                       -1   
1500                     1                   -2                        0   
1522                     1                   -1                        0   
1545                     1                   -1                        0   
1560                     0                    0                        0   
1642                     0                    0                       -1   
1681                     1                   -1                        0   
1705                     0                    0                        0   

      Anti-Gay Laws  Homosexuality Illegal  Pride Banned  Locals Hostile  \
1390              0                      0             0               0   
1408              0                      0             0              -1   
1428              0                      0            -1              -1   
1467              0                     -1             0              -1   
1500             -1                     -1             0              -1   
1522              0                      0             0               0   
1545              0                      0             0              -1   
1560              0                      0            -1              -1   
1642              0                      0             0               0   
1681              0                      0             0              -1   
1705              0                      0            -1              -1   

      Prosecution  Murders  Death Sentence  Trans Gender Rights  \
1390            0        0               0                  NaN   
1408            0       -2               0                  NaN   
1428            0        0               0                  NaN   
1467            0        0               0                  NaN   
1500           -2       -2              -5                  NaN   
1522            0        0 

In [13]:
# Correct totals
df['Total'] = df[factors].sum(1)

In [14]:
# Check totals back
sum(df.Total != df[factors].sum(1))

0

In [15]:
df[factors].describe()

Antidiscrimination Legislation  Marriage / Civil Partnership  \
count                           2e+03                         2e+03   
mean                            8e-01                         3e-01   
std                             1e+00                         9e-01   
min                            -1e+00                        -1e+00   
25%                             0e+00                         0e+00   
50%                             0e+00                         0e+00   
75%                             2e+00                         1e+00   
max                             3e+00                         2e+00   

       Adoption Allowed  LGBT Marketing  Equal Age of Consent  \
count             2e+03           1e+03                 2e+03   
mean              3e-01           1e-01                 4e-01   
std               7e-01           4e-01                 6e-01   
min              -1e+00           0e+00                -1e+00   
25%               0e+00           0e+00                 0e+00   
50%               0e+00           0e+00                 0e+00   
75%               0e+00           0e+00                 1e+00   
max               2e+00           2e+00                 1e+00   

       Religious Influence  HIV Travel Restrictions  Anti-Gay Laws  \
count                2e+03                    2e+03          2e+03   
mean                -9e-01                   -2e-01         -5e-01   
std                  8e-01                    5e-01          5e-01   
min                 -2e+00                   -3e+00         -2e+00   
25%                 -2e+00                    0e+00         -1e+00   
50%                 -1e+00                    0e+00          0e+00   
75%                  0e+00                    0e+00          0e+00   
max                  1e+00                    1e+00          0e+00   

       Homosexuality Illegal  Pride Banned  Locals Hostile  Prosecution  \
count                  2e+03         2e+03           2e+03        2e+03   
mean                  -3e-01        -6e-02          -1e+00       -3e-01   
std                    5e-01         3e-01           7e-01        5e-01   
min                   -2e+00        -2e+00          -2e+00       -2e+00   
25%                   -1e+00         0e+00          -2e+00        0e+00   
50%                    0e+00         0e+00          -1e+00        0e+00   
75%                    0e+00         0e+00          -1e+00        0e+00   
max                    0e+00         1e+00           1e+00        0e+00   

       Murders  Death Sentence  Trans Gender Rights  Intersex / 3rd Option  \
count    2e+03           2e+03                8e+02                  4e+02   
mean    -3e-01          -1e-01                7e-02                 -8e-01   
std      7e-01           7e-01                1e+00                  6e-01   
min     -2e+00          -5e+00               -2e+00                 -1e+00   
25%      0e+00           0e+00               -1e+00                 -1e+00   
50%      0e+00           0e+00                0e+00                 -1e+00   
75%      0e+00           0e+00                1e+00                 -1e+00   
max      0e+00           0e+00                3e+00                  2e+00   

       Conversion Therapy  
count               4e+02  
mean               -9e-01  
std                 5e-01  
min                -2e+00  
25%                -1e+00  
50%                -1e+00  
75%                -1e+00  
max                 1e+00

### Dataset Structure / Features of Interest / Assesment

There are 1,777 entries in the dataset that list the Country ranking per Total Score in period 2012-2021.

It features year, country, ranking, total (score), and 14 Human Rights categories. These categories indicates the level of acceptance of these Rights within a country, and range from marriage equality, anti-discrimination and sodomy laws, pride parade bans to hate crimes like death penalty.

Year and country are the only categorial variables, the rest of them are numeric.
Total is the summatory of the 14 Human Rights categories.

----------------------------------------------

- __Country__:                         1777 non-null   object / Change to category / 3 invalid: Aceh, Chechnia, Macao (extraneous) / Inconsistent country names
- __Ranking__:                         1777 non-null   int64
- __Year__:                            1777 non-null   int64  / Change to date
- __Total__:                           1777 non-null   int64  
- __Antidiscrimination Legislation__:  1771 non-null   float64 / Missing 6: Bonaire and Guam 2012-2014
- __Marriage / Civil Partnership__:    1774 non-null   float64 / Missing 3: Guam 2012-2014
- __Adoption Allowed__:                1774 non-null   float64 / Missing 3: Guam 2012-2014
- __LGBT Marketing__:                  1394 non-null   float64 / Missing 383: All 2018 & all 2019
- __Equal Age of Consent__:            1777 non-null   int8   
- __Religious Influence__:             1771 non-null   float64 / Missing 6: Bonaire and Gibraltar 2012-2014
- __HIV Travel Restrictions__:         1753 non-null   float64 / Missing 24: 8 Caribbean islands 2012-2014
- __Anti-Gay Laws__:                   1774 non-null   float64 / Missing 3: Bonaire 2012-2014
- __Homosexuality Illegal__:           1777 non-null   int8   
- __Pride Banned__:                    1777 non-null   int8   
- __Locals Hostile__:                  1774 non-null   float64 / Missing 3: Gibraltar 2012-2014
- __Prosecution__:                     1777 non-null   int8   
- __Murders__:                         1777 non-null   int8   
- __Death Sentence__:                  1777 non-null   int8   
- __Trans Gender Rights__:             787 non-null    float64 / Missing 990 since included since 2018 
- __Intersex / 3rd Option__:           404 non-null    float64 / Missing 1373 since included since 2020
- __Conversion Therapy__:              404 non-null    float64 / Missing 1373 since included since 2020

Recalculate ranking since invalid countries were deleted and totals were corrected

In [16]:
# Sort dataframe by Year, Total and Country
df.sort_values(by = ['Year', 'Total', 'Country'], ascending = [1,0,1], inplace=True)

# Reset index of dataframe (from 1 instead of 0)
df.index = np.arange(1, len(df) + 1)

In [17]:
# Assign SPARTACUS annual ranking per total (for Curvy Bump Slope Chart)

Period = df.Year.unique() # [2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021]

for yr in Period: # for e/year

    # year subset
    df_per_year = df[df.Year == yr]
    
    # value counts of Total
    val_ct = df_per_year.Total.value_counts().sort_index(ascending = 0)
    
    rank = 1
    for total in val_ct.index:
        
        # assign previous ranking plus count of previous ranking
        df.loc[(df.Year == yr) & (df.Total == int(total)), 'Ranking'] = rank
        rank += val_ct[total]

In [18]:
df.head(15)

Country  Ranking  Year  Total  Antidiscrimination Legislation  \
1           Sweden        1  2012      9                               3   
2          Belgium        2  2012      8                               2   
3      Netherlands        2  2012      8                               2   
4          Iceland        4  2012      7                               2   
5           Canada        5  2012      6                               2   
6          Denmark        5  2012      6                               1   
7           Norway        5  2012      6                               1   
8            Spain        5  2012      6                               2   
9   United Kingdom        5  2012      6                               2   
10         Uruguay        5  2012      6                               2   
11          France       11  2012      5                               2   
12         Germany       11  2012      5                               1   
13    Saint Martin       11  2012      5                               2   
14       Argentina       14  2012      4                               0   
15         Austria       14  2012      4                               1   

    Marriage / Civil Partnership  Adoption Allowed  LGBT Marketing  \
1                              2                 2               1   
2                              2                 2               1   
3                              2                 2               1   
4                              2                 2               0   
5                              2                 2               1   
6                              1                 2               1   
7                              2                 2               0   
8                              2                 2               1   
9                              1                 2               1   
10                             1                 2               0   
11                             1                 0               1   
12                             1                 1               1   
13                             2                 2               0   
14                             2                 2               1   
15                             1                 0               1   

    Equal Age of Consent  Religious Influence  HIV Travel Restrictions  \
1                      1                    0                        0   
2                      1                    0                        0   
3                      1                    0                        0   
4                      1                    0                        0   
5                      0                    0                       -1   
6                      1                    0                        0   
7                      1                    0                        0   
8                      1                   -1                        0   
9                      1                    0                        0   
10                     1                    0                        0   
11                     1                    0                        0   
12                     1                    0                        0   
13                     1                   -1                        0   
14                     1                   -1                        0   
15                     1                    0                        0   

    Anti-Gay Laws  Homosexuality Illegal  Pride Banned  Locals Hostile  \
1               0                      0             0               0   
2               0                      0             0               0   
3               0                      0             0               0   
4               0                      0             0               0   
5               0                      0             0               0   
6               0                      0    

In [19]:
print(df.loc[137:140, 'Ranking']) # 2012 - 2013 
print(df.loc[275:278, 'Ranking']) # 2013 - 2014
print(df.loc[413:416, 'Ranking']) # 2014 - 2015
print(df.loc[601:604, 'Ranking']) # 2015 - 2016
print(df.loc[793:796, 'Ranking']) # 2016 - 2017
print(df.loc[986:989, 'Ranking']) # 2017 - 2018
print(df.loc[1170:1173, 'Ranking']) # 2018 - 2019
print(df.loc[1364:1367, 'Ranking']) # 2019 - 2020
print(df.loc[1563:1566, 'Ranking']) # 2020 - 2021

137    137
138    138
139      1
140      2
Name: Ranking, dtype: int64
275    137
276    138
277      1
278      2
Name: Ranking, dtype: int64
413    137
414    138
415      1
416      1
Name: Ranking, dtype: int64
601    186
602    188
603      1
604      1
Name: Ranking, dtype: int64
793    190
794    192
795      1
796      1
Name: Ranking, dtype: int64
986    191
987    193
988      1
989      1
Name: Ranking, dtype: int64
1170    182
1171    184
1172      1
1173      1
Name: Ranking, dtype: int64
1364    192
1365    194
1366      1
1367      1
Name: Ranking, dtype: int64
1563    197
1564    199
1565      1
1566      2
Name: Ranking, dtype: int64


### Reshape the dataframe for Visualization

In [20]:
# Melt Categories for Visualization

# Duplicate Total column for future sorting
df['Total_Score'] = df['Total']

id_v  = ['Country', 'Ranking', 'Year', 'Total_Score']
val_v = factors.insert(0, 'Total')

df = pd.melt(df, id_vars=id_v, value_vars=val_v, var_name='Category', value_name='Measure')


In [21]:
# For e/category (except for Total) reassing new ranking

# Sort dataframe by Year and Total
df.sort_values(by = ['Year', 'Ranking', 'Country'], inplace=True)

# Reset index of dataframe (from 1 instead of 0)
df.index = np.arange(1, len(df) + 1)

# Create column New_Ranking with value 0
df.insert(0,'New_Ranking', 0)

for yr in Period: # for e/year
    
    # Since there are categories that are Nan for some years
    # assign max of unique countries (to avoid any ranking in 0) per year to all categories except Total
    df.loc[(df.Year == yr) & (df.Category != 'Total'), 'New_Ranking'] = len( df[df.Year == yr].Country.unique() )
    df.loc[(df.Year == yr) & (df.Category == 'Total'), 'New_Ranking'] = df.Ranking

    for f in factors:
        
        # year and category subset
        df_yr_ft = df[(df.Year == yr) & (df.Category == f)]
        
        df_yr_ft.sort_values(['Measure', 'Ranking', 'Country'], ascending = [0, 1, 1], inplace = True) 
        
        # New column with tuples of Ranking and Measure will be the parameter to rank
        # first by Measure and then by Ranking and Country, in this order
        df_yr_ft['Rank_Measure'] = list(zip(df_yr_ft.Ranking, df_yr_ft.Measure))
        
        
        # Value counts of tuples (Rank, Measure)
        val_ct = df_yr_ft.Rank_Measure.value_counts()
        
        
        rank = 1
        for uni_tup in df_yr_ft.Rank_Measure.unique():
            
            df.loc[(df.Year == yr) & (df.Category == f) \
                    & (df.Ranking == uni_tup[0]) & (df.Measure == uni_tup[1]), 'New_Ranking'] = rank

            rank += val_ct[uni_tup]


df.sort_values(by = ['Year', 'New_Ranking', 'Country'], inplace=True)

In [22]:
# Saving to excel

# Resort columns for Template
df2 = df[['Country', 'Category', 'Year', 'New_Ranking']] # Dimension, Category, Time, Rank, Measure // ActualTime

# Rename columns for Template
df2.rename(columns={'Country': 'Dimension', 'Year': 'Time', 'New_Ranking': 'Rank'}, inplace=True)

# Add last column Year as 'ActualTime' for Template
df2['ActualTime'] = df.Year

df2.to_excel(r'D:\Tableau\GTI_2012-2021_Spartacus.xlsx') # for curvy visual with categories